File: mobile-rn/src/state/activeRunStore.ts
import { secureStore } from '../services/secureStore';
import type { RunRequest } from '../models/run';

/**
 * Persists the in-flight run handle so the app can restore a streaming run on
 * relaunch — the RN equivalent of the web app's `sessionStorage` +
 * `tryRestoreActiveRun()` and the Swift `ActiveRunStore`. On launch the app
 * reads this, calls `GET /api/runs/{id}`, and reconnects the SSE stream.
 *
 * The handle is tiny and non-secret; it is stored via `expo-secure-store` simply
 * to avoid adding an extra persistence dependency for the MVP.
 */
const KEY = 'stockagents.activeRun';

export interface ActiveRun {
  runId: string;
  request: RunRequest;
  nodes: string[];
}

export const activeRunStore = {
  async save(run: ActiveRun): Promise<void> {
    await secureStore.set(KEY, JSON.stringify(run));
  },
  async load(): Promise<ActiveRun | null> {
    const raw = await secureStore.get(KEY);
    if (!raw) return null;
    try {
      return JSON.parse(raw) as ActiveRun;
    } catch {
      return null;
    }
  },
  async clear(): Promise<void> {
    await secureStore.delete(KEY);
  },
};

File: ios/StockAgents/Services/ActiveRunStore.swift
import Foundation

/// Persists the in-flight run handle so the app can restore a streaming run on
/// relaunch — the native equivalent of the web app's `sessionStorage` +
/// `tryRestoreActiveRun()`. On launch the app reads this, calls
/// `GET /api/runs/{id}`, and reconnects the SSE stream (plan §3.2).
struct ActiveRunStore {
    private static let key = "stockagents.activeRun"
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// What we need to fully reconstruct + reconnect a run after relaunch.
    struct ActiveRun: Codable, Equatable {
        let runID: String
        let request: RunRequest
        let nodes: [String]
    }

    func save(_ run: ActiveRun) {
        guard let data = try? JSONEncoder().encode(run) else { return }
        defaults.set(data, forKey: ActiveRunStore.key)
    }

    func load() -> ActiveRun? {
        guard let data = defaults.data(forKey: ActiveRunStore.key) else { return nil }
        return try? JSONDecoder().decode(ActiveRun.self, from: data)
    }

    func clear() {
        defaults.removeObject(forKey: ActiveRunStore.key)
    }
}
